File: src/board.ts
import { Cell } from "./cell";

export class Board {
  private cells: Cell[];
  private active: boolean = true;

  constructor(private boardSize: number) {
    this.cells = new Array(boardSize * boardSize).fill(Cell.Empty);
  }

  setCell(x: number, y: number, cell: Cell) {
    this.cells[x + y * this.boardSize] = cell;
  }

  getCell(x: number, y: number) {
    return this.cells[x + y * this.boardSize];
  }

  isActive() {
    return this.active;
  }

  setActive(active: boolean) {
    this.active = active;
  }

  isRowFilled(row: number, player: Cell) {
    for (let i = 0; i < this.boardSize; i++) {
      if (this.getCell(i, row) !== player) {
        return false;
      }
    }
    return true;
  }

  isColumnFilled(column: number, player: Cell) {
    for (let i = 0; i < this.boardSize; i++) {
      if (this.getCell(column, i) !== player) {
        return false;
      }
    }
    return true;
  }

  isDiagonalFilled(player: Cell) {
    for (let i = 0; i < this.boardSize; i++) {
      if (this.getCell(i, i) !== player) {
        return false;
      }
    }
    return true;
  }

  isAntiDiagonalFilled(player: Cell) {
    for (let i = 0; i < this.boardSize; i++) {
      if (this.getCell(i, this.boardSize - i - 1) !== player) {
        return false;
      }
    }
    return true;
  }

  isDone() {
    for (let i = 0; i < this.boardSize; i++) {
      if (
        this.isRowFilled(i, Cell.X) ||
        this.isColumnFilled(i, Cell.X) ||
        this.isDiagonalFilled(Cell.X) ||
        this.isAntiDiagonalFilled(Cell.X)
      ) {
        return true;
      }
      if (
        this.isRowFilled(i, Cell.O) ||
        this.isColumnFilled(i, Cell.O) ||
        this.isDiagonalFilled(Cell.O) ||
        this.isAntiDiagonalFilled(Cell.O)
      ) {
        return true;
      }
    }
    return false;
  }

  getWinner() {
    for (let i = 0; i < this.boardSize; i++) {
      if (
        this.isRowFilled(i, Cell.X) ||
        this.isColumnFilled(i, Cell.X) ||
        this.isDiagonalFilled(Cell.X) ||
        this.isAntiDiagonalFilled(Cell.X)
      ) {
        return Cell.X;
      }
      if (
        this.isRowFilled(i, Cell.O) ||
        this.isColumnFilled(i, Cell.O) ||
        this.isDiagonalFilled(Cell.O) ||
        this.isAntiDiagonalFilled(Cell.O)
      ) {
        return Cell.O;
      }
    }
    return null;
  }

  isDraw() {
    for (let i = 0; i < this.boardSize; i++) {
      for (let j = 0; j < this.boardSize; j++) {
        if (this.getCell(i, j) === Cell.Empty) {
          return false;
        }
      }
    }
    return true;
  }
}

File: super-ttt-server/lib/super_ttt/logic/board.dart
import 'package:super_ttt_server/super_ttt/logic/cell.dart';

class Board {
  List<Cell> cells = [];
  bool active = true;
  int boardSize;

  Board({this.boardSize = 3}) {
    for (var i = 0; i < boardSize * boardSize; i++) {
      cells.add(Cell.empty);
    }
  }

  setCell(int x, int y, Cell cell) {
    cells[x + y * boardSize] = cell;
  }

  getCell(int x, int y) {
    return cells[x + y * boardSize];
  }

  isActive() {
    return active;
  }

  setActive(bool active) {
    this.active = active;
  }

  isRowFilled(int row, Cell cellType) {
    for (var i = 0; i < boardSize; i++) {
      if (getCell(i, row) != cellType) {
        return false;
      }
    }
    return true;
  }

  isColumnFilled(int column, Cell cellType) {
    for (var i = 0; i < boardSize; i++) {
      if (getCell(column, i) != cellType) {
        return false;
      }
    }
    return true;
  }

  isDiagonalFilled(Cell player) {
    for (var i = 0; i < boardSize; i++) {
      if (getCell(i, i) != player) {
        return false;
      }
    }
    return true;
  }

  isAntiDiagonalFilled(Cell player) {
    for (var i = 0; i < boardSize; i++) {
      if (getCell(i, boardSize - i - 1) != player) {
        return false;
      }
    }
    return true;
  }

  isDone() {
    for (var i = 0; i < boardSize; i++) {
      if (isRowFilled(i, Cell.x) ||
          isColumnFilled(i, Cell.x) ||
          isDiagonalFilled(Cell.x) ||
          isAntiDiagonalFilled(Cell.x)) {
        return true;
      }
      if (isRowFilled(i, Cell.o) ||
          isColumnFilled(i, Cell.o) ||
          isDiagonalFilled(Cell.o) ||
          isAntiDiagonalFilled(Cell.o)) {
        return true;
      }
    }
    return false;
  }

  getWinner() {
    for (var i = 0; i < boardSize; i++) {
      if (isRowFilled(i, Cell.x) ||
          isColumnFilled(i, Cell.x) ||
          isDiagonalFilled(Cell.x) ||
          isAntiDiagonalFilled(Cell.x)) {
        return Cell.x;
      }
      if (isRowFilled(i, Cell.o) ||
          isColumnFilled(i, Cell.o) ||
          isDiagonalFilled(Cell.o) ||
          isAntiDiagonalFilled(Cell.o)) {
        return Cell.o;
      }
    }
    return null;
  }

  isDraw() {
    for (var i = 0; i < boardSize; i++) {
      for (var j = 0; j < boardSize; j++) {
        if (getCell(i, j) == Cell.empty) {
          return false;
        }
      }
    }
    return true;
  }

  Map<String, dynamic> toJson() {
    return {
      'cells': cells.map((e) => e.stringify()).toList(),
      'active': active,
    };
  }
}
